### ISSUE TITLE
Do not rely on THREE global

### ISSUE THREAD

**Issue by arjanfrans** [Jam3/three-bmfont-text#13]
I am building my application using Webpack and started using ES6 modules to import three.js functions. Since this module relies on the THREE global, I ran into a problem that it is undefined inside this module.

It works for now by setting the global THREE manually. However, I think it would be nice to add three.js as a peer dependency on this module and import the required function from three.js inside, instead of relying on the global.

What do you think?

**Comment by mattdesl** [comment]
This already has a peer dependency on ThreeJS, it's just not explicit in the `package.json`. :smile: This means you can use any build of ThreeJS (i.e. with browserify/webpack, or just a plain <script> tag before your bundle).

Explicitly requiring the `three` package will cause a lot of issues. Many production-scale projects ship with their own versions of ThreeJS rather than the `three` module. Also, because there are so many different versions of ThreeJS, it's very likely that you will end up with multiple instances in the same bundle if your code depends on a module that depends on some mismatched version of ThreeJS.

See more here:
https://github.com/mattdesl/three-orbit-controls/issues/17#issuecomment-233139027
https://gist.github.com/mattdesl/06d056f10322532e6e07bc97ef91b557

**Comment by iangilman** [comment]
I ran across this issue as well, using webpack, and I had to dig around to find the webpack command to make it work. I'm posting it here in case other people had the same problem. In your webpack.config.js configuration object, add:

```
  plugins: [
    new webpack.ProvidePlugin({
      THREE: 'three'
    })
  ]
```

**Comment by mflux** [comment]
Running into this issue as well. We have a build that uses webpack build that loads pre-bundled three from a lib folder. Including bmfont-text via npm will cause THREE to not be found, both using global or window object because it's all loaded at once in the bundle. Even using ProvidePlugin solution above does not work since THREE isn't installed via NPM.

Not exactly sure what the right solution for this lib is though :|

**Comment by iangilman** [comment]
Indeed... seems like one solution would be to refactor this library to not alias THREE until it's first called. That wouldn't change the API for anyone else, but would allow for workarounds for scenarios like @mflux has.

**Comment by jamieowen** [comment]
I made these quick changes to get this library working with ES6 / Browserify.

https://github.com/jamieowen/three-bmfont-text/commit/10e3ecc38d326498a90591b4028cfad39afb4285

It's not really PR ready as i had to pull in some other deps code directly, but just thought i'd reference it here..
But i've had similar nightmares with ES6 and duplicate versions of three.js. 

The cleanest and safest way i've discovered so far for ES6 use for custom three modules is to reference three.js only as a proper peerDependency in the package.json.   Similar to this ( https://github.com/jamieowen/three-material-modifier/blob/master/package.json ) 

But definitely _don't_ reference three in dev or prod dependencies ( even for demo-making purposes ) as a version still gets put in the deps node_modules.

Also you need to add NODE_PATH=./node_modules before your browserify/budo related commands otherwise node will complain about not finding three in your deps package.
```
NODE_PATH=./node_modules budo src/index.js
```

I've had some weird shit happen when other dependencies include three.js, like declaring reusable Geometry in the top level scope of a js file causes some weird rendering problems that are basically impossible to debug as there are no errors, just nothing appears but only in certain circumstances.. confusing as hell! :)

I haven't figured out why those problems happen exactly but this setup has really helped me.

**Comment by AlbertoElias** [comment]
This does make it difficult to use with ES2015 modules though, as THREE needs to be set as a global, as a hack, somewhere before this module is imported, but due to the dependency graph, it's not that easy to do properly

**Comment by pablokvitca** [comment]
Has anyone gotten the library working inside Angular 5? Declaring THREE as a global variable (window.THREE = THREE or document.THREE = THREE) don't seem to do anything. I keep getting an error complaining THREE is not defined. I think a simple way to allow developers to make the library use whatever version they want of three.js is for the THREE object to be passed as an argument when calling the library. This approach would allow maintaining the flexibility of the library while also supporting modules on ES2015 and frameworks such as Angular and others.

**Comment by dbuck** [comment]
I went the slightly different and hackier route. I don't want to add ALL of 'three' to our bundle size (we've got some stripping going on), but we can forward the parts needed into the global three to allow the package to work without modifications.

```
// three-bmfont-shim.ts
import { BufferGeometry, Sphere, Box3, Texture, Color, BufferAttribute } from 'three';
(window as any).THREE = { ...(window as any).THREE, BufferGeometry, Sphere, Box3, Texture, Color, BufferAttribute };
```

**Comment by stur86** [comment]
I agree this is a pretty bad decision. There are hacks to go around it but that's all they are, hacks. I don't see what's the problem with importing THREE in the library's own file locally, since anyway that will default to whatever the version any given user has installed in `node_modules` is.

**Comment by wilson0x4d** [comment]
this could be better.

**Comment by mattdesl** [comment]
🤷‍♂️

ThreeJS has come a long way in terms of npm/modularity/build tools since this issue was opened (in 2016), so it makes sense to re-evaluate, and that discussion has already started in a PR [here](https://github.com/Jam3/three-bmfont-text/pull/17#issuecomment-552480588). But it's not really as simple as just accepting the PR as it stands now and making a breaking change, as it will likely break a lot of the software that is using this module in production.

For example, AFrame is using this module, and (no surprise) they have forked ThreeJS and are using their own version in their library rather than relying on node_modules. So if I do include `require('three')` in this module, then any user installing AFrame would end up with two versions of ThreeJS in node_modules. Forking ThreeJS like this is really common in larger productions. I do agree the best thing for the future is to depend directly on `three` module and just let the dependents figure it out themselves (for example aliasing `three` imports to their own fork), but it's not really a decision I take lightly and I don't see any urgency to it since both choices here lead to headache and more issues in my inbox, and I also have other things I'd rather be doing with my time than worrying about whether or not THREE is in global scope in this module. An even better solution would be to pass this torch to a team that has a more vested interest in the longevity of this module (such as somebody from AFrame), but a lot of that team disappeared when Mozilla had their layoffs.

So yeah; the decision may appear poor now in hindsight but it's also more nuanced than it might appear, and times have change since 2016.

 🍻 

p.s. @wilson0x4d Commenting "this could be better" is a waste of my time as it goes to my inbox as an email. Maybe just fork this module and move on, or don't use my module at all if you don't agree with it.

**Comment by stur86** [comment]
To me it sounds like it would be on whoever is curating _that_ library to then take care that, if they fork THREE, they also fork modules like this one and make the necessary changes. It leaves me puzzled that this is the one case that controls the decision rather than the most straightforward one where I `npm install` THREE, then this module, and expect them to work out of the box like one would think they ought to. But I understand that at this point it might be problematic to just backtrack, as many of those libraries would just end up not being fixed because the maintainers might never realise what's changed, etc. Either way, personally I actually ended up forking this module into my own version just for this.

**Comment by SvetlozarValchev** [comment]
Still not resolved? No one holds three.js as a global variable since the last century. Why was it decided this way and why isn't it fixed yet?

**Comment by JustalK** [comment]
I just run into this problem.

I fork the repository and added the line :
```
const THREE = require('three')
```

Or you could just add it globally as said in the previous comment
